spmd/tensor/parallel/api.py

```diff
@@ -86,3 +87,144 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
+    module: nn.Linear,
+    device_mesh: DeviceMesh,
+    weight_distribute_spec: Sequence[Placement],
+    bias_distribute_spec: Sequence[Placement],
+) -> None:
+    """
+    This util function parallelize the weight and bias of a nn.Linear
+    and replace parameter in place.
+
+    Args:
+        module (nn.Linear):
+            :class:``nn.Linear`` object to be parallelized.
+        device_mesh (DeviceMesh):
+            :class:``DeviceMesh`` object which contains
+            how we distribute tensor across GPUs.
+        weight_distribute_spec (Sequence[Placement]):
+            the tensor distribute spec of DTensor for weight.
+        bias_distribute_spec (Sequence[Placement]):
+            the tensor distribute spec of DTensor for bias.
+
+    Return:
+        None
+    """
+    weight = nn.Parameter(
+        distribute_tensor(module.weight, device_mesh, weight_distribute_spec)
+    )
+    bias = nn.Parameter(
+        distribute_tensor(module.bias, device_mesh, bias_distribute_spec)
+    )
+    module.register_parameter("weight", weight)
+    module.register_parameter("bias", bias)
+
+
+def _parallelize_mlp(
+    module: nn.Module,
+    device_mesh: DeviceMesh,
+    parallel_style: ParallelStyle = PairwiseParallel(),
+    tp_mesh_dim: int = 0,
+) -> None:
+    """
+    This function assumes the input module is a sequence of nn.Linear
+    and we parallelize the module based on the given parallel style.
+    We don't change the FQN of each sub-module and replace each parameter
+    in place.
+
+    Args:
+        module (nn.Module):
+            :class:``nn.Module`` object to be parallelized.
+        device_mesh (DeviceMesh):
+            :class:``DeviceMesh`` object which contains
+            how we distribute tensor across GPUs.
+        parallel_style (ParallelStyle):
+            :class:``ParallelStyle`` object which contains how
+            we prepare input/output for Tensor Parallelism.
+        tp_mesh_dim (int):
+            the dimension of ``device_mesh`` where we perform
+            Tensor Parallelism on.
+
+    Return:
+        None
+
+    .. warning::
+        We now only support ``PairwiseParallel`` for now.
+    """
+
+    # Define hook functions needed for preparing Input/Output.
+    def _module_forward_pre_hook(*args):  # pyre-ignore[2, 3]
+        return args[0](args[-1][0], *args[1:-2])
+
+    def _module_forward_hook(*args):  # pyre-ignore[2, 3]
+        return args[0](args[-1], *args[1:-3])
+
+    if not isinstance(parallel_style, PairwiseParallel):
+        raise NotImplementedError(
+            "Only support PairwiseParallel for MLP parallelization."
+        )
+
+    if not _has_even_num_linears(module):
+        raise RuntimeError("We only support even number of Linear for MLP.")
+
+    if device_mesh.ndim > 1:
+        device_mesh = _create_1d_device_mesh(device_mesh, tp_mesh_dim)
+
+    for i, m in enumerate(
+        filter(lambda x: isinstance(x, nn.Linear), module.children())
```

Comment:
I'm wondering if we should ensure sth weird not happen: i.e. `linear0 -> emb -> linear1` will be treated as a legit one but it's actually not, maybe we should check the layers in the middle and ensure they don't have parameters?

Reply:
But softmax and layerNorm also have parameters. It's hard to differentiate it from embedding in a generic way.